### DISCUSSION TITLE
List of supported HMAC.

### DISCUSSION THREAD

**Discussion by stefbon** [openssl/openssl#21774] [Q&A] [answered]
Hi,

I'm looking for the supported HMAC's, and when reading manpages, I start with EVP_MAC, where is mentioned: "See ...(MAC) in OSSL_PROVIDER-default(7)" for the list of algorithms supported and the syntax (esp. the names).
When going there I see under the header "Message Authentication Code (MAC)" for HMAC see EVP_MAC-HMAC(7).  When going there, also no names of supported algo's. Where can I find them?

S. Bon

**Comment by t8m** [comment]
Any supported MD algorithm is supported with HMAC (at least in the HMAC implementation of the default provider).

**Reply by mattcaswell** [reply]
> Any supported MD algorithm is supported with HMAC (at least in the HMAC implementation of the default provider).

There are a handful of exceptions, i.e. you cannot use SHAKE-128 or SHAKE-256.

**Reply by t8m** [reply]
That's for the FIPS provider only, isn't it?

**Reply by stefbon** [reply]
Thanks. Where can I find a list of supported MD algorithms? And what is the syntax to use a HMAC (==the name to be used in EVP_MAC_fetch)?

**Reply by mattcaswell** [reply]
No, it applies to the default provider too:

https://github.com/openssl/openssl/blob/79cdbe893da0c613db97356d05c0b088e885707f/crypto/hmac/hmac.c#L45-L50

**Reply by t8m** [reply]
Ah, ok

**Reply by mattcaswell** [reply]
> Thanks. Where can I find a list of supported MD algorithms? And what is the syntax to use a HMAC (==the name to be used in EVP_MAC_fetch)?

It depends which providers you have loaded. The list of digests available in the default provider is listed on this page:

https://www.openssl.org/docs/man3.1/man7/OSSL_PROVIDER-default.html

When you call `EVP_MAC_fetch` you just specify "HMAC" as the name. When you call `EVP_MAC_init` you supply the name of the digest in the `OSSL_MAC_PARAM_DIGEST` param.

**Comment by t8m** [comment]
Or in other words - the HMAC implementation in the default provider can use any EVP_MD implementation even from other, third party providers. So it is not possible to list all algorithms supported in HMAC.

**Comment by t8m** [comment]
> Thanks. Where can I find a list of supported MD algorithms? And what is the syntax to use a HMAC (==the name to be used in EVP_MAC_fetch)?

Supported hash algorithms in the default provider are in OSSL_PROVIDER-default(7).

You fetch the `"HMAC"` algorithm. Then you specify the hash name as a parameter.

**Comment by mattcaswell** [comment, marked as answer]
There is an example of how to do this here:

https://github.com/openssl/openssl/blob/master/demos/mac/hmac-sha512.c

**Comment by stefbon** [comment]
So I cannot use the name HMAC_SHA256 directly in fetch?

**Comment by mattcaswell** [comment]
> So I cannot use the name HMAC_SHA256 directly in fetch?

No.

**Comment by stefbon** [comment]
I come from libgcrypt where the syntax is different. The example you provided is very helpfull. I have to get used to the construction of the params used there. I have to write a wrapper to handle this I think.

**Comment by stefbon** [comment]
Thanks both for answering. My question is solved.
